### PR TITLE
Support Byron Epoch boundary blocks

### PR DESCRIPTION
**Pull request by nc6**
This PR introduces the following:

- Updates to cardano-ledger. This means changing a few things which have been updated in that repo, such as the naming for the proxy delegation certificates. The updates are needed to expose functions for dealing with serialisation of epoch boundary blocks.
- Introduce a protocol extension `WithEBBs`. The reason for this abstraction is that it still makes more sense to associate the "raw" `ByronBlock` with PBFT. If instead we used `ByronBlockOrEBB`, everything would have to have associated `Nothing` cases where the block turned out to be an EBB. Instead, we keep the association of PBFT with the raw `ByronBlock` and associate `ByronBlockOrEBB` with the extension `WithEBBs PBFT`.

In spite of retaining `ByronBlock` for this reason, in most cases everything defined over `ByronBlock` is now defined over `ByronBlockOrEBB`.

This is still in draft, since there's an assertion failure in `AnchoredFragment` which I don't yet understand.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import Data.Proxy (Proxy(..))
+import Ouroboros.Consensus.Protocol.Abstract
+
+class ( SupportedHeader p (HOEHdr hoe)
+      -- , HasHeader (HOEEbb hoe)
+      ) => HeaderSupportsWithEBB p hoe where
+
+  type HOEHdr hoe :: *
+  type HOEEbb hoe :: *
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs]: Hmm, these names are quite cryptic. We won't use them too much, so longer names are not too annoying.

What about `HeaderOfEBB` and `HeaderOfBlock`. Or, if we only import this module qualified as `WithEBBs`, then even `EBB` and `(Regular)Block` might work.

It'd also use `h` instead of `hoe`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  newtype NodeConfig   (WithEBBs p) = WithEBBNodeConfig {unWithEBBNodeConfig :: NodeConfig p}
+
+  preferCandidate       (WithEBBNodeConfig cfg) = preferCandidate       cfg
+  compareCandidates     (WithEBBNodeConfig cfg) = compareCandidates     cfg
+  checkIsLeader         (WithEBBNodeConfig cfg) = checkIsLeader         cfg
+  rewindChainState      (WithEBBNodeConfig cfg) = rewindChainState      cfg
+  protocolSecurityParam (WithEBBNodeConfig cfg) = protocolSecurityParam cfg
+
+  applyChainState (WithEBBNodeConfig cfg) lv b cs = case
+    eitherHeaderOrEbb (Proxy :: Proxy p) b of
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs]: Can we use `cfg` instead of `Proxy :: Proxy p` here?

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  type HOEEbb hoe :: *
+
+  eitherHeaderOrEbb :: proxy p -> hoe -> Either (HOEEbb hoe) (HOEHdr hoe)
+
+-- | Extension of protocol @p@ to support addition 'epoch boundary blocks'; that
+-- is, blocks which do not contribute to the general running of the protocol,
+-- but do add to the hash chain.
+--
+-- This is entirely a legacy concern, and should go away after the Byron era.
+data WithEBBs p
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs]: Oh, I like that this is just a protocol extension/modifier :+1:

File: ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs [hunk 4]
```diff
@@ -32,24 +33,75 @@ import           Ouroboros.Consensus.Ledger.Byron
...
+      . annotateBoundary given
+      $ boundaryData
+  where
+    boundaryData = CC.Block.BoundaryValidationData
+                   boundaryLength
+                   (Left pbftGenesisHash)
+                   epoch
+                   (CC.Common.ChainDifficulty 0)
+                   ()
+    boundaryLength = 0 -- Since this is a demo and we ignore the length, set this
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs]: What is the `boundaryLength`?

File: ouroboros-consensus/ouroboros-consensus.cabal [hunk 5]
```diff
@@ -93,6 +93,7 @@ library
                        Ouroboros.Consensus.Protocol.Praos
                        Ouroboros.Consensus.Protocol.Signed
                        Ouroboros.Consensus.TxSubmission
+                       Ouroboros.Consensus.Protocol.WithEBBs
```
[the remark this answers is not in the document]
Comment (nc6) [reply ouroboros-consensus/ouroboros-consensus.cabal]: Oh, sorry, think this is a rebase artifact.

File: ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs (hunk 4 above)
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs]: Length of the boundary block

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs]: Of course :man_facepalming:

**Comment by mrBliss** [comment]
This depends on [cardano-ledger#566](https://github.com/input-output-hk/cardano-ledger/pull/566), which still needs to be merged.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 6]
```diff
@@ -83,6 +93,14 @@ import qualified Ouroboros.Consensus.Util.SlotBounded as SB
   Byron blocks and headers
 -------------------------------------------------------------------------------}
 
+-- | Unfortunate Byron requires some information that is morally static but
+-- actually comes from configuration files on startup. We use reflection to
+-- provide this to type class instances.
+type ByronGiven = (
+    Given Crypto.ProtocolMagicId
+  , Given CC.Slot.EpochSlots
+  )
```
Comment (avieth) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: It was done this way in `cardano-sl`, as an improvement over the original design that used template haskell. Both are less than ideal: with TH, each build is fixed to one protocol magic; with reflections, it can be changed by runtime configuration, but it's not at all clear whether one process is capable of using two different protocol magics.

**Comment by nc6** [comment]
Rebased atop master, addressed a bunch of the comments.

File: stack.yaml [hunk 7]
```diff
@@ -47,6 +52,7 @@ extra-deps:
     commit: e8bfc1294e088f90e5ae0b4aedbc82ee46ac5ee4
 
   - bimap-0.4.0
+  - binary-0.8.7.0
```
Comment (mrBliss) [inline stack.yaml]: Why is this needed?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 8]
```diff
@@ -515,61 +378,214 @@ instance Condense (Header (ByronBlock cfg)) where
...
         $ hdr
 
-instance Condense (ChainHash (ByronBlock cfg)) where
+{-------------------------------------------------------------------------------
+  Epoch Boundary Blocks
+-------------------------------------------------------------------------------}
+
+newtype ByronBlockOrEBB cfg = ByronBlockOrEBB
+  { unByronBlockOrEBB :: CC.Block.ABlockOrBoundary ByteString
+  } deriving (Eq, Show)
```
Comment (avieth) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: For byron-proxy, I'll need a type which includes Byron blocks and EBBs, but which also can be serialised. Since there's no way to serialise an `ABlockOrBoundary ByteString`, what I've done there is use

```Haskell
type Block = Annotated (ABlockOrBoundary ByteString) ByteString
```

as the block type, and encoded it using CBOR tag 24 so that when I decode it from the wire I can give the top-level annotation (I have the bytes).

Now I'm working on using `ChainDB` in byron-proxy, so I'll face the same problem with the block type chosen to go in the `ChainDB`. I'll need the top-level annotation there as well. Perhaps we could change `ByronBlockOrEBB` to include this?

File: stack.yaml (hunk 7 above)
Comment (nc6) [reply stack.yaml]: Whenever I don't have it, I can't build. Not sure why nobody else seems to have this problem...

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 6 above)
Comment (erikd) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Would be considered to be old fashioned to suggest that we just use parameter passing for this?

File: ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs [hunk 9]
```diff
@@ -32,24 +33,75 @@ import           Ouroboros.Consensus.Ledger.Byron
...
+     )
+  => NodeConfig ByronEBBExtNodeConfig
+  -> SlotNo                       -- ^ Current slot
+  -> BlockNo                      -- ^ Current block number
+  -> ChainHash (ByronBlockOrEBB cfg)   -- ^ Previous hash
+  -> [GenTx (ByronBlockOrEBB cfg)]     -- ^ Txs to add in the block
+  -> ()                           -- ^ Leader proof ('IsLeader')
+  -> m (ByronBlockOrEBB ByronDemoConfig)
+forgeBlockOrEBB cfg curSlot curNo prevHash txs () = case prevHash of
+  GenesisHash -> forgeGenesisEBB cfg curSlot
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs]: This means that only the very first block (the one after genesis) produced will be an EBB. I thought that an EBB was produced at the start of each epoch?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs [hunk 10]
```diff
@@ -625,12 +641,174 @@ annotateHeader epochSlots =
...
+
+  Since we will not be creating further boundary blocks, these utilities do not
+  exist in the cardano-ledger repo, but we need them for the genesis case in the
+  demo.
+  -------------------------------------------------------------------------------}
+
+annotateBoundary :: Crypto.ProtocolMagicId
+                 -> CC.Block.BoundaryValidationData ()
+                 -> CC.Block.BoundaryValidationData ByteString
+annotateBoundary pm =
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: While debugging the demo with `--real-pbft` (#705): I discovered that there must be a bug somewhere in this function (or in cardano-ledger#566): if I give it an EBB with `boundaryPrevHash = Left _`, the block I get back after `annotateBoundary` suddenly has `boundaryPrevHash = Right _`. If I replace `annotateBoundary` with `fmap (const mempty)`, then `boundaryPrevHash` remains `Left _`. This is the reason why the first produced block doesn't fit onto the chain.

Can you add a round-trip QC property for `fromCBOR.. . toCBOR..` + find the actual bug? :slightly_smiling_face:

File: stack.yaml (hunk 7 above)
Comment (mrBliss) [reply stack.yaml]: Aarrgh, this line has cost me soooo much time rebuilding when switching branches: stack always has to build +100 dependencies whenever the version of `binary` changes in stack.yaml. Why can't stack remember that it has built these dependencies before? (I'm not blaming you, just venting :slightly_smiling_face: )

File: ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs (hunk 9 above)
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Demo/Ledger/Byron/Forge.hs]: Yes, but this isn't needed for the demo. We only need a block at the actual genesis to distinguish the genesis hash.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 6 above)
Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: That'd be my ideal solution. But since the design uses the typeclass `HasHeader`, something that makes these parameters _seem_ static is needed (otherwise the instances couldn't be implemented).

Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: Something more like an ML module would be nice here. Then the particular values of protocol magic and epoch slots could determine a new type for a byron block.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs (hunk 8 above)
Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: This is also important for using the `ChainDB` to serve a `Byron` peer. That requires coming up with legacy `cardano-sl` `Block` and `BlockHeader` values, but if I'm using a `ChainDB` that uses new `cardano-ledger` variants, I can't do that without having the bytes from which they were decoded.

Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: I don't understand this comment - why can't you serialise an `ABlockOrBoundary ByteString`? It's already annotated, and we have explicit serialisers for it (`toCBORABOBBlock` and `toCBORABOBBoundary`, but we could easily wrap these together).

Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: `ABlockOrBoundary ByteString` couldn't be serialized, but now can since you recently merged a pull request into `cardano-ledger`.

Comment (avieth) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs]: For `BoundaryValidationData a` the annotation `a` is for the header bytes only, so `ABOBBoundary (BoundaryValidationData ByteString)` doesn't have the bytes for the whole block. That's why I put the annotation out front. Looks like the `toCBORABOBBoundary` just makes up a body... I'm not sure if that's actually ok. The Byron peers might not accept it.

File: stack.yaml (hunk 7 above)
Comment (avieth) [reply stack.yaml]: Use nix :)

Comment (mrBliss) [reply stack.yaml]: I would like to, but using `cabal new-build` in a nix-shell is still recompiling my dependencies, which is defeating the point :man_facepalming:

File: ouroboros-network/src/Ouroboros/Network/Chain.hs [hunk 11]
```diff
@@ -123,7 +123,10 @@ validExtension
 validExtension c b = blockInvariant b
                   && headHash c == blockPrevHash b
                   && headSlot c <  blockSlot b
-                  && headBlockNo c == pred (blockNo b)
+                  -- TODO This is a hack which should be addressed through 'WithOrigin'
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/Chain.hs]: @avieth when can we expect the sequel of #706? :slightly_smiling_face:  We need `WithOrigin` for `BlockNo` (in addition to `SlotNo`) too, as the genesis EBB has `BlockNo 0`, the empty chain cannot have `BlockNo 0` too.

If you're busy with the byron-proxy, I could take a stab at it instead.

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Chain.hs]: I did this, for ouroboros-network only https://github.com/input-output-hk/ouroboros-network/pull/774

It would probably be best if you did the follow up for ouroboros-consensus. I found there were some non-obvious choices to be made in there.

**Comment by mrBliss** [comment]
Sorry @nc6, but #727 caused some merge conflicts. It's mostly moving/splitting/renaming files, but some identifiers were renamed too.
